### PR TITLE
Fix zoomus.conf and get Wayland working

### PR DESCRIPTION
**Pull request by TheEvilSkeleton**
Closes https://github.com/flathub/us.zoom.Zoom/issues/22
Closes https://github.com/flathub/us.zoom.Zoom/pull/182
Closes https://github.com/flathub/us.zoom.Zoom/pull/143

### CODE REVIEW

**Comment by TheEvilSkeleton** [comment]
Works as it should in my end. This is good to be merged

**Comment by TheEvilSkeleton** [comment]
@5n00p4eg hope you don't mind the ping, but can you test if [this](https://github.com/flathub/us.zoom.Zoom/pull/217/#issuecomment-701795941) Flatpak works for you?

Just keep in mind that the Wayland screensharing will work the second time your launch Zoom (https://github.com/flathub/us.zoom.Zoom/pull/219#issuecomment-701425156).

**Comment by physkets** [comment]
I installed the modified flatpak as you suggested, but it reports an issue with the symlink:
```
$ flatpak run us.zoom.Zoom//test 
ln: failed to create symbolic link '/home/kitchu/.var/app/us.zoom.Zoom/.config/zoomus.conf': File exists
```
Also, attempting to screen share gives the same dialog as in https://github.com/flathub/us.zoom.Zoom/issues/22#issuecomment-700509678

**Comment by TheEvilSkeleton** [comment]
> I installed the modified flatpak as you suggested, but it reports an issue with the symlink:
> 
> ```
> $ flatpak run us.zoom.Zoom//test 
> ln: failed to create symbolic link '/home/kitchu/.var/app/us.zoom.Zoom/.config/zoomus.conf': File exists
> ```

This should be fine because it launches (right?). The symlink has already been set, so it will output an error that the file already exists (which isn't a problem)

As for the actual issue, I'll try to find a fix for it. Zoom's a pain in the ass to work with

**Comment by TheEvilSkeleton** [comment]
@physkets for https://github.com/flathub/us.zoom.Zoom/issues/22#issuecomment-700509678, can you try running:

```
flatpak run --env=XDG_CURRENT_DESKTOP=gnome us.zoom.Zoom//test
```

and tell me if it worked or not?

**Comment by TheEvilSkeleton** [comment]
```bash
$ echo $XDG_CURRENT_DESKTOP 
GNOME
```
Maybe you can try running

```
flatpak run --env=XDG_CURRENT_DESKTOP=GNOME us.zoom.Zoom//test
```

instead?

**Comment by physkets** [comment]
> `flatpak run --env=XDG_CURRENT_DESKTOP=gnome us.zoom.Zoom//test`

Looks like that changes things. I mean the usual error message does not appear, and the option to share does not show the warning.

But I'm still not able to actually share. I'm guessing maybe 'xdg-desktop-portal-gtk' needs to be installed for that?

**Comment by TheEvilSkeleton** [comment]
> But I'm still not able to actually share. I'm guessing maybe `xdg-desktop-portal-gtk` needs to be installed for that? Or does that already exist by default? I have `xdg-desktop-portal-kde` from my distro's packages; is that what is needed in the Flatpak too? Or something else like Pipewire?

I honestly have no idea.

I don't know how to make it install `xdg-desktop-portal-gtk` *yet*, but I'll try to figure it out.

**Comment by TheEvilSkeleton** [comment]
@physkets can you try to run

```
flatpak run --socket=session-bus --env=XDG_CURRENT_DESKTOP=GNOME us.zoom.Zoom//test
```
and report back to me?

**Comment by physkets** [comment]
> `flatpak run --socket=session-bus --env=XDG_CURRENT_DESKTOP=GNOME us.zoom.Zoom//test`

Same as before, it still doesn't explicitly object, but there is still no option to share the desktop:

Screenshot_20201001_221814

This is unlike in X11, where there are two options in the 'Basic' section: A whiteboard and a desktop share.

**Comment by TheEvilSkeleton** [comment]
I see. I wish I was able to test it myself, but I don't have anybody to test with currently. Is there a way I can share my display in Zoom without creating a new account?

Also, have you tried using the version that came in your package manager? If so, does the problem still occur or does it actually let you screen share?

**Comment by physkets** [comment]
> I see. I wish I was able to test it myself, but I don't have anybody to test with currently.

I've been testing it using two different machines.

>  Is there a way I can share my display in Zoom alone and without creating a new account?

Well, you can start a new meeting, and then click the 'share screen' button. That will pop-up these options, and then you will know what it allows.

> Also, have you tried using the version that came in your package manager?

I did once, but I think I used it by switching to the X11 session, because, it isn't very reliable on Wayland (https://github.com/flathub/us.zoom.Zoom/issues/215) for me.

**Comment by TheEvilSkeleton** [comment]
> Well, you can start a new meeting, and then click the 'share screen' button. That will pop-up these options, and then you will know what it allows. You don't really need another account/person to join that call.

I checked multiple videos on how to start a new meeting, and they all require to sign in/up. I ended up signing up with a temp mail.

File: us.zoom.Zoom.json [hunk 1]
```diff
@@ -49,6 +50,7 @@
                     "type": "script",
                     "dest-filename": "zoom.sh",
                     "commands": [
+                        "ln -s ~/.var/app/us.zoom.Zoom/config/zoomus.conf ~/.var/app/us.zoom.Zoom/.config",
```
Comment (barthalion) [inline us.zoom.Zoom.json:53]: I'm tired, so just to make sure I understand it right: despite `--persist=.config`, Zoom looks for zoomus.conf elsewhere, ignoring XDG_SOMETHING var?

Comment (TheEvilSkeleton) [reply us.zoom.Zoom.json:53]: Yes and no. Zoom looks in `~/.config`, but the problem is that the user still has to manually copy/paste or create a `zoomus.conf` because `--persist=.config` doesn't create it for you. The symlink is used to symlink `~/.var/app/us.zoom.Zoom/config/zoomus.conf to `~/.var/app/us.zoom.Zoom/.config`.

In case you still don't understand: the Zoom Flatpak looks inside `~/.var/app/us.zoom.Zoom/.config`, but for some reason `--persist=.config` still doesn't map `zoomus.conf` to `~/.var/app/us.zoom.Zoom/.config/zoomus.conf`, so the symlink is used to symlink that file to where the Zoom Flatpak looks.

But yes, Zoom doesn't follow XDG standards. It just hardcodes to the directories which is why the Zoom Flatpak is pain to get it to work.

**Comment by TheEvilSkeleton** [comment]
So overriding `filesystem=host` worked for me...

**Comment by TheEvilSkeleton** [comment]
How can one tell if it's successfully screen recording or not? Also, is it screen recording?

Screenshot from 2020-10-01 15-43-59

**Comment by TheEvilSkeleton** [comment]
How can one tell if it's actually screen sharing or not? Also, is it successfully screen sharing?

Screenshot from 2020-10-01 15-43-59

**Comment by TheEvilSkeleton** [comment]
So this is a weird bug(?). Moving the `.zoom` directory from `~/.var/app/us.zoom.Zoom/.zoom` to `~/.zoom`, and then symlinking them [fixes(???)](https://user-images.githubusercontent.com/50847364/94856982-5109e680-0420-11eb-8c19-0e1243e69e08.png) the issue.

**Comment by TheEvilSkeleton** [comment]
@physkets I just tried on Manjaro Plasma (Wayland), and it looks like the screen sharing option isn't showing up.

**Comment by TheEvilSkeleton** [comment]
I also tried adding `--env=XDG_SESSION_DESKTOP=gnome` but that didn't work either. I really don't know what to do at this point.

**Comment by StoneToken** [comment]
> Build [29050 successful](https://flathub.org/builds/#/builders/32/builds/29050)
> To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/27971/us.zoom.Zoom.flatpakref
> ```

Hi! Tried on Fedora 32 Wayland, screen sharing can be launched, but the shared screen is black.

**Comment by physkets** [comment]
Adding home or host filesystem access does not help, but in doing so, I noticed that I had to login again. This seems to be because Zoom defaults to using its own config directory on the desktop. So maybe access to `xdg-desktop/.zoom` should be added?

**Comment by TheEvilSkeleton** [comment]
> > Build [29050 successful](https://flathub.org/builds/#/builders/32/builds/29050)
> > To test this build, install it from the testing repository:
> > ```
> > flatpak install --user https://dl.flathub.org/build-repo/27971/us.zoom.Zoom.flatpakref
> > ```
> 
> Hi! Tried on Fedora 32 Wayland, screen sharing can be launched, but the shared screen is black.

Can you try running

```
ln -s ~/.var/app/us.zoom.Zoom/.zoom ~/
```

and tell me if it works or not? (Assuming `~/.zoom` doesn't exist.)

**Comment by TheEvilSkeleton** [comment]
> Adding home or host filesystem access does not help, but in doing so, I noticed that I had to login again. This seems to be because Zoom defaults to using its own config directory on the desktop. So maybe access to `xdg-desktop/.zoom` should be added?

That is due to ~~Linux being Linux~~ Zoom's configuration files to go to `~/.zoom` and `~/.config/zoomus.conf` instead of staying inside of `~/.var/app/us.zoom.Zoom`. You should be able to just move `~/.var/app/us.zoom.Zoom/.zoom` to `~/.zoom`, and you won't need to login again.

**Comment by physkets** [comment]
> `ln -s ~/.var/app/us.zoom.Zoom/.zoom ~/`

`ln: failed to create symbolic link '/home/user/.zoom': File exists`

> `xdg-desktop/.zoom` doesn't need to be added because allowing home/host will already do that. Maybe not using `--persist=.zoom` and allowing `~/.zoom` might fix it, but I'll need others to confirm it for me.

Well, a lot of people (including me) rescind blanket access via `override --nofilesystem=home`, so is such cases, specifying exactly what it needs is always better.

**Comment by TheEvilSkeleton** [comment]
> > `ln -s ~/.var/app/us.zoom.Zoom/.zoom ~/`
> 
> `ln: failed to create symbolic link '/home/user/.zoom': File exists`

You will have to move your `~/.zoom` folder to someone else. But keep in mind that this message was directed to @StoneToken because he would have the same issue as me for screen sharing. Running that command (I think) fixed the issue on my side.

> Well, a lot of people (including me) rescind blanket access via override --nofilesystem=home, so is such cases, specifying exactly what it needs is always better.

In this case, we will have to change all the permissions for that (which is a terrible idea in my opinion, unless it has to be done).

If you want to do that, then you must allow rw access to the following directories:

- `~/Documents/Zoom` (not `xdg-documents/Zoom` because Zoom hardcodes)
- `~/.zoom
- `~/.config/zoomus.conf`

You should be able to restrict everything else.

**Comment by TheEvilSkeleton** [comment]
I made the Flatpak have rw access to `~/.zoom` because `--persist=.zoom` just wouldn't work with screen sharing.

@StoneToken the sharing worked for me when using https://github.com/flathub/us.zoom.Zoom/pull/217/#issuecomment-702731244. Is it the same for you?

**Comment by TheEvilSkeleton** [comment]
I had to remove `--persist=.zoom` because Zoom was able to look at my screen when not using `--persist=.zoom`.

**Comment by StoneToken** [comment]
> > > Build [29050 successful](https://flathub.org/builds/#/builders/32/builds/29050)
> > > To test this build, install it from the testing repository:
> > > ```
> > > flatpak install --user https://dl.flathub.org/build-repo/27971/us.zoom.Zoom.flatpakref
> > > ```
> > 
> > 
> > Hi! Tried on Fedora 32 Wayland, screen sharing can be launched, but the shared screen is black.
> 
> Can you try running (before launching Zoom)
> 
> ```
> ln -s ~/.var/app/us.zoom.Zoom/.zoom ~/
> ```
> 
> and tell me if the screen sharing works or not? (Assuming `~/.zoom` doesn't exist.) I think (but not sure) it worked for me.

Hi! Making a symlink worked all right! Screen sharing works on https://dl.flathub.org/build-repo/27971/us.zoom.Zoom.flatpakref

**Comment by TheEvilSkeleton** [comment]
@StoneToken can you try removing `~/.zoom` and `~/.var/app/us.zoom.Zoom/` directories, and then try [this Flatpak](https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-702731244) and tell me if the Flatpak works out of the box? If the first time doesn't work, can you try starting the program once or twice more?

It might not work the first time you launch Zoom, because of (assuming) this: https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-701807264

**Comment by StoneToken** [comment]
> @StoneToken can you try removing `~/.zoom` and `~/.var/app/us.zoom.Zoom/` directories, and then try [this Flatpak](https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-702731244) and tell me if the screen sharing works out of the box? If the first launch doesn't work, can you try launching the program once or twice more?
> 
> It might not work the first time you launch Zoom, because of (assuming) this: [#217 (comment)](https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-701807264).
> 
> [This Flatpak](https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-702731244) worked for me out of the box.

Well, steps I did:
1. remove all zoom clients
2. rm -rf ~/.zoom
rm -rf ~/.var/app/us.zoom.Zoom
3. flatpak install --user https://dl.flathub.org/build-repo/28035/us.zoom.Zoom.flatpakref
4. flatpak run us.zoom.Zoom
5. Fist time it wrote
sed: can't read /home/user/.var/app/us.zoom.Zoom/config/zoomus.conf: No such file or directory
6. Fist time it showed my webcam as screen sharing (sorry didn't make screen), after stopping cam video - it showed none screen sharing options except white board
7. Kill zoom, start again
8. Start up log:
`flatpak run us.zoom.Zoom 
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/Unknown Organization': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/glib-2.0': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/pulse': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/zoomus.conf': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/fontconfig': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/mesa_shader_cache': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qt_compose_cache_little_endian_stonetoken': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtshadercache': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8-lockfile': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/tmp': File exists
ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/zoom': File exists`
9. Screen sharing seemed to work all right
10. Meeting recording complained about incorrect audio settings and recorded black screen instead of screen sharing

**Comment by TheEvilSkeleton** [comment]
> 5. Fist time it wrote
> sed: can't read /home/user/.var/app/us.zoom.Zoom/config/zoomus.conf: No such file or directory
> 8.  Start up log:
> `flatpak run us.zoom.Zoom ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/Unknown Organization': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/glib-2.0': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/pulse': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/zoomus.conf': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/fontconfig': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/mesa_shader_cache': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qt_compose_cache_little_endian_stonetoken': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtshadercache': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8-lockfile': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/tmp': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/zoom': File exists`

Those errors are normal because, for 5., `/home/user/.var/app/us.zoom.Zoom/config/zoomus.conf` won't exist the first time you launch so it'll output this error, and for 8., those symlinks already have been created which is why it outputs those errors.

> 10. Meeting recording complained about incorrect audio settings and recorded black screen instead of screen sharing

Does this still happen when you allow `filesystem=home`? That sounds more like an upstream issue.

**Comment by TheEvilSkeleton** [comment]
@jadahl, I'm confirming that it doesn't work.

Using https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-702731244:
Screenshot from 2020-10-04 15-19-24

Using https://github.com/flathub/us.zoom.Zoom/pull/217#issuecomment-703301843:
Screenshot from 2020-10-04 15-18-11

**Comment by TheEvilSkeleton** [comment]
@jadahl looks like https://github.com/flathub/us.zoom.Zoom/pull/217/commits/d5e3d214607376d996e398826113bafefab97250 works for me. I'll keep that commit.

**Comment by StoneToken** [comment]
> > 1. Fist time it wrote
> >    sed: can't read /home/user/.var/app/us.zoom.Zoom/config/zoomus.conf: No such file or directory
> 
> `/home/user/.var/app/us.zoom.Zoom/config/zoomus.conf` won't exist the first time you launch so it'll output this error, which is normal. The second time you launch should do it correctly.
> 
> > 1. Start up log:
> >    `flatpak run us.zoom.Zoom ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/Unknown Organization': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/glib-2.0': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/pulse': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.config/zoomus.conf': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/fontconfig': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/mesa_shader_cache': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qt_compose_cache_little_endian_stonetoken': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtshadercache': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/qtsingleapp-zoom-3e8-lockfile': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/tmp': File exists ln: failed to create symbolic link '/home/user/.var/app/us.zoom.Zoom/.cache/zoom': File exists`
> 
> Those symlinks already have been created which is why it outputs those errors, which is normal.
> 
> > 1. Meeting recording complained about incorrect audio settings and recorded black screen instead of screen sharing
> 
> Does this still happen when you allow `filesystem=home`? That sounds more like an upstream issue.

For the first time screen sharing works OK (no matter --filesystem=home)
For the next time screen sharing stops working (the same was last time, I was not attentive enough)

Tried on flatpak install --user https://dl.flathub.org/build-repo/28230/us.zoom.Zoom.flatpakref

**Comment by TheEvilSkeleton** [comment]
>  For the next time screen sharing stops working (the same was last time, I was not attentive enough)

That's really weird. I honestly don't know what the problem is.

> (the same was last time, I was not attentive enough)

What do you mean by that?

**Comment by StoneToken** [comment]
> > For the next time screen sharing stops working (the same was last time, I was not attentive enough)
> 
> That's really weird. I honestly don't know what the problem is.
> 
> > (the same was last time, I was not attentive enough)
> 
> What do you mean by that?

When I tested previous build behavior was exactly the same. I just didn't try to test zoom several times.

**Comment by TheEvilSkeleton** [comment]
> > > For the next time screen sharing stops working (the same was last time, I was not attentive enough)
> > 
> > 
> > That's really weird. I honestly don't know what the problem is.
> > > (the same was last time, I was not attentive enough)
> > 
> > 
> > What do you mean by that?
> 
> When I tested previous build behavior was exactly the same. I just didn't try to test zoom several times.

So if I understood it right, launching Zoom for the very first time worked, but the second launch and after stopped working?

**Comment by TheEvilSkeleton** [comment]
> Build [29420 successful](https://flathub.org/builds/#/builders/32/builds/29420)
> To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
> ```

Works well for me.

**Comment by rosesproctor** [comment]
Did I type something wrong?

`flatpak install --user https://dl/flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
error: Can't load uri https://dl/flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref: Error resolving “dl”: Temporary failure in name resolution
`

**Comment by rosesproctor** [comment]
Thank you so much for this fix! It's working for me as well <3

**Comment by TheEvilSkeleton** [comment]
I'll ready it because it works fine for some of us.

**Comment by TheEvilSkeleton** [comment]
@physkets Just confirming, Zoom [hard depends on GNOME](https://github.com/TheEvilSkeleton/us.zoom.Zoom/blob/fix_zoomus/us.zoom.Zoom.json#L18) for screen sharing. There's not much we can do apart from switching to GNOME (by using a VM, installing it in bare metal

**Comment by JJoaoxx** [comment]
@TheEvilSkeleton , many thanks for working on this. Newbie question here, I've installed the build but can't seem to be able to launch it, the usual commands are coming back as not recognized. Sorry if this is a stupid question...

**Comment by TheEvilSkeleton** [comment]
> @TheEvilSkeleton , many thanks for working on this. Newbie question here, I've installed the build but can't seem to be able to launch it, the usual commands are coming back as not recognized. Sorry if this is a stupid question...

Can you run

```
flatpak run us.zoom.Zoom
```

and send me the output?

**Comment by JJoaoxx** [comment]
@TheEvilSkeleton , I appreciate your response. I will have to fight you on on this one and insist it was, indeed, a stupid question, here's my argument: following your response, I now know that flatpak apps need the command 'flatpak run' rather than just the name of the application alone. There's nothing wrong with the build and it ran perfectly after that! So see, I win! JK, many thanks again!

**Comment by TheEvilSkeleton** [comment]
Newbie questions don't mean they're stupid :P We all learn

**Comment by TheEvilSkeleton** [comment]
# TL;DR of this MR

zoomus.conf works, the sandboxing is still the same, except that `~/.var/app/us.zoom.Zoom/.zoom` has been moved to `~/.zoom` due to it breaking screen sharing in GNOME + Wayland.

We have discovered that Zoom hard depends on GNOME for screen sharing in Wayland, so we cannot do anything for those that run Wayland on another DE/WM.

This MR is good to go. We got screen sharing working on GNOME Wayland.

**Comment by StoneToken** [comment]
> Build [29420 successful](https://flathub.org/builds/#/builders/32/builds/29420)
> To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
> ```

Hi! Screen sharing works all right! But screen recording still records blank black screen.

**Comment by TheEvilSkeleton** [comment]
> > Build [29420 successful](https://flathub.org/builds/#/builders/32/builds/29420)
> > To test this build, install it from the testing repository:
> > ```
> > flatpak install --user https://dl.flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
> > ```
> 
> Hi! Screen sharing works all right! But screen recording still records blank black screen.

```bash
[fedora@localhost ~]$ flatpak override --show --user us.zoom.Zoom 
[Context]
sockets=cups;pcsc;ssh-auth;system-bus;session-bus
filesystems=home;host-etc;host-os;host

[System Bus Policy]
org.mate.*=own
org.kde.*=own
org.freedesktop.*=own
org.gnome.*=own

[Session Bus Policy]
org.gnome.*=own
org.freedesktop.*=own
org.kde.*=talk
org.mate.*=own
org.gnome.Shell=none
```

I tried. I really tried, but I just can't get it to work. I really don't know what Zoom needs to screen record. If someone knows how to get it to work, please let me know.

This MR is going to stay as is.

**Comment by TheEvilSkeleton** [comment]
> > Build [29420 successful](https://flathub.org/builds/#/builders/32/builds/29420)
> > To test this build, install it from the testing repository:
> > ```
> > flatpak install --user https://dl.flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
> > ```
> 
> Hi! Screen sharing works all right! But screen recording still records blank black screen.

```bash
[fedora@localhost ~]$ flatpak override --show --user us.zoom.Zoom 
[Context]
sockets=cups;pcsc;ssh-auth;system-bus;session-bus
filesystems=host-etc;host-os;host

[System Bus Policy]
org.mate.*=own
org.kde.*=own
org.freedesktop.*=own
org.gnome.*=own

[Session Bus Policy]
org.gnome.*=own
org.freedesktop.*=own
org.kde.*=talk
org.mate.*=own
org.gnome.Shell=none
```

I tried. I really tried, but I just can't get it to work. I really don't know what Zoom needs to screen record. If someone knows how to get it to work, please let me know.

This MR is going to stay as is.

**Comment by StoneToken** [comment]
> > > Build [29420 successful](https://flathub.org/builds/#/builders/32/builds/29420)
> > > To test this build, install it from the testing repository:
> > > ```
> > > flatpak install --user https://dl.flathub.org/build-repo/28333/us.zoom.Zoom.flatpakref
> > > ```
> > 
> > 
> > Hi! Screen sharing works all right! But screen recording still records blank black screen.
> 
> ```shell
> [fedora@localhost ~]$ flatpak override --show --user us.zoom.Zoom 
> [Context]
> sockets=cups;pcsc;ssh-auth;system-bus;session-bus
> filesystems=host-etc;host-os;host
> 
> [System Bus Policy]
> org.mate.*=own
> org.kde.*=own
> org.freedesktop.*=own
> org.gnome.*=own
> 
> [Session Bus Policy]
> org.gnome.*=own
> org.freedesktop.*=own
> org.kde.*=talk
> org.mate.*=own
> org.gnome.Shell=none
> ```
> 
> I tried. I really tried, but I just can't get it to work. I really don't know what Zoom needs to screen record. If someone knows how to get it to work, please let me know.
> 
> I don't think I'll be committing further. Everything (except screen recording) works well now.

Thank you very much for your effort!

File: us.zoom.Zoom.json [hunk 2]
```diff
@@ -49,6 +47,8 @@
                     "type": "script",
                     "dest-filename": "zoom.sh",
                     "commands": [
+                        "ln -s ~/.var/app/us.zoom.Zoom/config/* ~/.var/app/us.zoom.Zoom/.config",
```
Comment (barthalion) [inline us.zoom.Zoom.json:50]: This really needs some `ifs` to check whether it already exists or not.

Comment (muelli) [reply us.zoom.Zoom.json:50]: I was thinking the very same. But I think it's relying on the directory being empty. In any case, at least a comment would be appropriate.  Because it looks like a gross hack :-/

Comment (TheEvilSkeleton) [reply us.zoom.Zoom.json:50]: > This really needs some `ifs` to check whether it already exists or not.

Can you submit an MR on the fork? It's my first time hearing about `ifs` and I have no idea what this means :(

Comment (BjoernDaase) [reply us.zoom.Zoom.json:50]: You are probably looking for https://stackoverflow.com/questions/51539586/how-do-i-use-the-if-then-else-condition-in-json-schema.

**Comment by BjoernDaase** [comment]
bot, build

**Comment by BjoernDaase** [comment]
I can confirm that this works perfectly fine under GNOME 3.38 on F33.

File: us.zoom.Zoom.json (hunk 2 above)
Comment (barthalion) [reply us.zoom.Zoom.json:50]: The problem here is that it makes an optimistic assumption of `~/.var/app/us.zoom.Zoom/config/` being empty and `~/.var/app/us.zoom.Zoom/.config` not existing. It's unlikely to be the case for all users after all changes since Zoom is on Flathub. It should check for edge cases and handle them gracefully instead of creating random symlinks.

Comment (TheEvilSkeleton) [reply us.zoom.Zoom.json:50]: @barthalion I just wrote a bash script:

```bash
#!/usr/bin/env bash

ZOOMUS=$HOME/.var/app/us.zoom.Zoom/.config/zoomus.conf
if [[ -f "$ZOOMUS" ]]; then
    :
else
    ln -s $HOME/.var/app/us.zoom.Zoom/config/* ~/.var/app/us.zoom.Zoom/.config
fi

ZOOMUS_WAYLAND=$HOME/.var/app/us.zoom.Zoom/.config/zoomus.conf
if [[ -f "$ZOOMUS_WAYLAND" ]]; then
    sed -i 's/enableWaylandShare\=false/enableWaylandShare\=true/' ~/.var/app/us.zoom.Zoom/config/zoomus.conf
else
    echo "File does not exist yet. Re-run Zoom to successfully symlink."
fi
```
but I really didn't understand the link you send, so I can't really do anything here.

**Comment by TheEvilSkeleton** [comment]
I know that the workaround is [ugly](https://github.com/flathub/us.zoom.Zoom/pull/217#discussion_r503129218), but due to my limitation in JSON and bash scripting, that's all I can provide. The workaround _works_, even though it's an ugly one.

**Comment by vilgotf** [comment]
Would be great if this could be merged to master and not be stuck on beta (since it seems to be working and this is a high priority issue)

**Comment by TheEvilSkeleton** [comment]
^ Agreed. Until then, we should try to keep improving it.

**Comment by muelli** [comment]
I won't merge this as is, as it is too messy for me to understand. Fortunately, there are other who could merge if they feel so inclined. I would merge to beta, though.

It looks to me like several independent issues that could be proposed separately.

**Comment by TheEvilSkeleton** [comment]
The main purpose of it was to merge it in beta, so we could improve it later.

File: us.zoom.Zoom.json (hunk 2 above)
Comment (TheEvilSkeleton) [reply us.zoom.Zoom.json]: Should be fixed now https://github.com/flathub/us.zoom.Zoom/pull/217/commits/ad0f11020b4d68048c6c8e74965e30a9f84ce0b8

**Comment by TheEvilSkeleton** [comment]
@muelli I commented the stuff so you can understand the script https://github.com/flathub/us.zoom.Zoom/pull/217/commits/a9b034b387b2535fc6b607b2e918b85e4c7078be. I tested it and it seems to be working now.

File: zoom.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+# Zoom hardcodes ~/.config instead of using XDG_CONFIG_HOME.
+ZOOMUS=$HOME/.var/app/us.zoom.Zoom/.config/zoomus.conf
```
Comment (muelli) [inline zoom.sh:4]: I thought we'd have a shorthand for that. `XDG_CONFIG` or so. Maybe I am mistaken, though.

**Comment by muelli** [comment]
As I've indicated, merging to beta is fine with me. But before that can go into production, the commit history needs to be cleaned up.
I suggest to break it up into smaller pieces.

**Comment by TheEvilSkeleton** [comment]
Sure, I'll try to clean everything up, and maybe also improve it.

**Comment by Funky185540** [comment]
Hi there,

Sorry for bumping a merged PR, but I really want to have that wayland fix for Zoom. I've been trying to get it installed on my PC, but it seems the artifacts from the flathubbot don't exist any more.

So I went to get Gnome Builder, but that tells me that "apply_extra" doesn't exist and fails the build process (Spoiler alert: I've never worked with Gnome Builder, nor have I ever built a flatpak application from sources). From what I see in the JSON config file it seems like "apply_extra" is in fact a script that grabs the Zoom archive and extracts it. Do I have to make something "special" (like build options or something?) to make it build properly? I even tried manually extracting the tar file to the /app/bin dir but that didn't have any effect either...

Any hints what exactly I'm doing wrong?

Steps I took:
- Clone an existing project in Gnome Builder
- Go to the project dir and checkout the "beta" branch
- Hit the "Build" button in Gnome Builder

P.S.: Thanks for (apparently) making Screen Sharing work!

**Comment by TheEvilSkeleton** [comment]
@Funky185540 have you tried using `flatpak-builder` instead? It's a CLI tool so you'll have to get it from the distro package manager.

From the root directory of the repo, run:

```bash
flatpak-builder --force-clean --install --install-deps-from=flathub --user build-dir us.zoom.Zoom.json
```

It should build it.

**Comment by Funky185540** [comment]
@TheEvilSkeleton 
Awesome, that worked! Thank you so much!
Would it be worthwhile maybe adding a README to the repo for anyone else interested in building it on their own? I could provide a PR for that, I guess.

And I can report from a quick test on my system (F33 workstation, running Wayland) that Screen Sharing now sort of works. Zoom seemingly can't share any apps/content that was already open/visible before the screen sharing started (Anything that was open, like e.g. other applications or the desktop background turns into grey noise for other viewers). I'll try to gather some more info on that. Is it worth already opening another Issue for this (Since technically the patch is still just in the "beta" branch)?

Again: Thanks for your effort!

**Comment by TheEvilSkeleton** [comment]
>  Would it be worthwhile maybe adding a README to the repo for anyone else interested in building it on their own? I could provide a PR for that, I guess.

I'm not sure if the maintainers are okay with it, but I support that idea. Flatpak lacks a lot of proper documentation; having some docs that will help users to do stuff on their own will make it better.

**Comment by barthalion** [comment]
Sorry but dropping random snippets into README files does not constitute a documentation.

@Funky185540 A package with these changes is available in Flathub beta. `flatpak remote-add --user flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo` and then install Zoom from there.

**Comment by rmader** [comment]
Concerning "Wayland support" I just wanted to drop that it may makes sense to point out somewhere that it is only expected to work on Gnome.

Zoom implements their desktop sharing on Wayland via the Gnome-private screenshot API, which is bad for performance, compatibility and, probably most importantly, security. The talk-name `org.gnome.Shell` gives Zoom full access to all Gnome APIs, effectively disabling the sandbox completely - it can run arbitrary code in the Wayland compositor.

Also, it would probably good to set the "sandbox level" accordingly so it's shown in the software center. Claiming that the app is sandboxed at all is quite misleading as long as access to `org.gnome.Shell` is present.
